File: flipped_material/python/code/guessing_game/guessing_game_9.py
from random import randint

def make_guess(min_value, max_value, moves):
    still_guessing = True
    while still_guessing:
        guess = int(input(f"Guess a number between {min_value} and {max_value} "))
        if guess > min_value or guess < max_value:
            still_guessing = False
    return guess, moves + 1

def auto_guess(low, high, last_guess, fitness):
    if fitness > 0:
        high = last_guess
    elif fitness < 0:
        low = last_guess
    guess = int((high+low)/2)
    return guess, low, high
    

def check_for_hints(guess, answer):
    if guess < answer:
        return "your guess was too low!"
    elif guess > answer:
        return "your guess was too high!"
    return f"{guess} was the right answer."

def guess_fitness_test(hint):
    if hint == "your guess was too low!":
        return -1
    elif hint == "your guess was too high!":
        return 1
    return 0


sum_of_moves = 0
games_played = 0
quickest_game = 999999
longest_game = -1
min_value = 1
max_value = 1000_000_000_000
for _ in range(0, 1000000):
    answer = randint(min_value, max_value)
    guess = -1
    moves = 0
    low_guess = min_value
    high_guess = max_value
    fitness = 0

    while answer != guess:
        guess, low_guess, high_guess = auto_guess(low_guess, high_guess, guess, fitness)
        moves += 1
        hint = check_for_hints(guess, answer)
        fitness = guess_fitness_test(hint)
    
    sum_of_moves += moves
    games_played += 1
    if moves < quickest_game: 
        quickest_game = moves
    if moves > longest_game:
        longest_game = moves
print(f"You played a total of {games_played:,} games")
print(f"On average, you took {sum_of_moves/games_played:,} moves")
print(f"The quickest game was {quickest_game}")
print(f"Your longest game was {longest_game}")

File: flipped_material/cpp/code/guessing_game/guessing_game9.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

int makeGuess(min_value, max_value){
   int stillGuessing = 1;
   int guess;
   while (stillGuessing) {
     printf("Pick a number between %d and %d\n", min_value, max_value);
     scanf("%d", &guess);
     if (guess >= min_value && guess <= max_value){
       stillGuessing = 0;
     }
   }
   return guess;
}

int fitnessTest(guess, answer){
  if (guess < answer) return -1;
  if (guess > answer) return 1;
  return 0;
}

void hint(int fitness, int guess, int answer, int low, int high) {
    if (fitness < 0) {
      printf("Your guess %d was too low %d!", guess, answer);
    } else if (fitness > 0) {
      printf("Your guess %d was too high %d!", guess, answer);
    } else {
      printf("You are the winner!");
    }
    printf(" %d %d\n", low, high);
}

int autoGuess(int min_value, int max_value){
  return (min_value+max_value+1)/2;
}


int main(void) {
  int min_value = 1;
  int max_value = 1000;
  int numberOfGames = 1000;
  int sum_of_moves = 0;
  int games_played = 0;
  int quickestGame = 9999;
  int longestGame = -1;
  srand(time(NULL));  
  for (int i = 0; i < numberOfGames; i++){
    
    int answer = rand() % (max_value-min_value+1)+min_value;
    int guess = -1;
    int fitness = 0;
    int moves = 0;
    int lowGuess = -1;
    int highGuess = max_value+1;

    while (answer != guess){
      if (fitness > 0){
        highGuess = guess; 
      } else if (fitness < 0) {
        lowGuess = guess;
      }
      guess = autoGuess(lowGuess, highGuess);
      moves++;
      fitness = fitnessTest(guess, answer);
      hint(fitness, guess, answer, lowGuess, highGuess);
    }
    sum_of_moves += moves;
    games_played++;
    if (moves > longestGame) longestGame = moves;
    if (moves < quickestGame) quickestGame = moves;
    
  }
  printf("You played a total of %d games\n", games_played);
  printf("On average, you took %d moves\n", sum_of_moves/games_played);
  printf("The quickest game was %d\n", quickestGame);
  printf("The longest game was %d\n", longestGame);
  return 0;
}